modules/setting/markup.go

```diff
@@ -25,42 +28,106 @@ type MarkupParser struct {
 	IsInputFile    bool
 }
 
+// MarkupSanitizer defines the policy for whitelisting attributes on
+// certain elements.
+type MarkupSanitizer struct {
+	Elements   []string
+	AllowAttrs []string
+	Regexps    []string
+}
+
 func newMarkup() {
-	extensionReg := regexp.MustCompile(`\.\w`)
 	for _, sec := range Cfg.Section("markup").ChildSections() {
 		name := strings.TrimPrefix(sec.Name(), "markup.")
 		if name == "" {
 			log.Warn("name is empty, markup " + sec.Name() + "ignored")
 			continue
 		}
 
-		extensions := sec.Key("FILE_EXTENSIONS").Strings(",")
-		var exts = make([]string, 0, len(extensions))
-		for _, extension := range extensions {
-			if !extensionReg.MatchString(extension) {
-				log.Warn(sec.Name() + " file extension " + extension + " is invalid. Extension ignored")
-			} else {
-				exts = append(exts, extension)
-			}
+		if name == "sanitizer" {
+			newMarkupSanitizer(name, sec)
+		} else {
+			newMarkupRenderer(name, sec)
 		}
+	}
+}
+
+func newMarkupSanitizer(name string, sec *ini.Section) {
+	// Implementation detail: go-ini/ini doesn't currently support
+	// ini arrays properly: https://github.com/go-ini/ini/issues/197
+
+	haveElement := sec.HasKey("ELEMENT")
+	haveAttr := sec.HasKey("ALLOW_ATTR")
+	haveRegexp := sec.HasKey("REGEXP")
+
+	if !haveElement && !haveAttr && !haveRegexp {
+		log.Warn("Skipping empty section: markup.%s.", name)
+		return
+	}
+
+	if !haveElement || !haveAttr || !haveRegexp {
+		log.Fatal("Missing required keys from markup.%s. Must have all three of ELEMENT, ALLOW_ATTR, and REGEXP defined!", name)
+	}
+
+	elements := sec.Key("ELEMENT").ValueWithShadows()
+	allowAttrs := sec.Key("ALLOW_ATTR").ValueWithShadows()
+	regexps := sec.Key("REGEXP").ValueWithShadows()
+
+	if len(elements) != len(allowAttrs) ||
+		len(elements) != len(regexps) ||
+		len(allowAttrs) != len(regexps) {
+		log.Fatal("All three keys in markup.%s (ELEMENT, ALLOW_ATTR, REGEXP) must be defined the same number of times! Got %d, %d, and %d respectively.", name, len(elements), len(allowAttrs), len(regexps))
+	}
 
-		if len(exts) == 0 {
-			log.Warn(sec.Name() + " file extension is empty, markup " + name + " ignored")
+	for index, pattern := range regexps {
+		if pattern == "" {
 			continue
 		}
 
-		command := sec.Key("RENDER_COMMAND").MustString("")
-		if command == "" {
-			log.Warn(" RENDER_COMMAND is empty, markup " + name + " ignored")
-			continue
+		// Validate when parsing the config that this is a valid regular
+		// expression. Then we can use regexp.MustCompile(...) later.
+		_, err := regexp.Compile(pattern)
+		if err != nil {
+			log.Fatal("In module.%s: REGEXP at definition %d failed to compile: %v", name, index+1, err)
```

Comment:
`log.Error()` is preferrable, since markup is not essential to start the instance (e.g. go to the admin menu).

Reply:
If the configuration supplied by the admin is bad (and they had to take some action to get to a bad configuration), isn't it preferable to error out completely so they can re-edit and fix it?

Otherwise the admin might not notice that their configuration was bad. In this case, I guess the result would be that all sanitization would be applied with no additional whitelisting, so the admin might eventually find out if someone complains. 

If you'd like me to change this still, I could.

Reply:
In the `settings` package, `Fatal()` is generally used for extreme situations; most settings will/should use their default values as a fallback mechanism. The admin might be about to install Gitea; we want them to be able to render the install page at least. 😉